### ISSUE TITLE
leading space in first line of git commit message breaks pipeline

### ISSUE THREAD

**Issue by ccmtaylor** [gocd/gocd#7774]
##### Issue Type

- Bug Report

##### Summary

A commit with leading whitespace in the first line, and a yaml-compatible non-indented list in the remaining lines breaks the commit message parsing in GoCD.  I suspect this was introduced in https://github.com/gocd/gocd/commit/45f63b08b664f8d0472e05117f5b15dd38e54395, and may be related to https://github.com/gocd/gocd/issues/7752, which also deals with git commit parsing and snakeYaml.

For example, this commit message produces the error below:

```
  empty commit
    
* with leading space in the first line
* and a list in the body
* breaks things
```

![image](https://user-images.githubusercontent.com/425325/75147429-8ec1c680-56fd-11ea-80a5-849c6ee65aaf.png)

I pasted the error from the server logs further down.

###### Basic environment details

* Go Version: `20.1.0 (11114-f2c1771e333ea92dae82cd78d2efa1310c1086f0)`
* JAVA Version: `13.0.1`
* OS: `Linux`

##### Steps to Reproduce

1. Create a pipeline you're willing to break, with a git material
2. make a commit as detailed above
```bash
git commit --allow-empty -m "  leading whitespace

* and
* a 
* list
"
```
3. `git push`
4. wait for the error

##### Expected Results

build pipeline should run as usual

##### Actual Results

The pipeline fails to run with the errors detailed above. Further commits don't help; the offending commit needs to be force-pushed away.

##### Possible Fix

I suspect this was introduced in https://github.com/gocd/gocd/commit/45f63b08b664f8d0472e05117f5b15dd38e54395, and may be related to https://github.com/gocd/gocd/issues/7752, which also deals with git commit parsing and snakeYaml.

##### Log snippets

```
2020-02-24 10:58:12,053 WARN  [237@MessageListener for MaterialUpdateListener] MaterialDatabaseUpdater:127 - [Material Update] Modification check failed for material: URL: git@github.com:soundcloud/fortune.git, Branch: master
org.yaml.snakeyaml.scanner.ScannerException: while scanning an alias
 in 'string', line 11, column 5:
        * with leading space in the firs ... 
        ^
expected alphabetic or numeric character, but found  (32)
 in 'string', line 11, column 6:
        * with leading space in the first ... 
         ^

        at org.yaml.snakeyaml.scanner.ScannerImpl.scanAnchor(ScannerImpl.java:1447)
        at org.yaml.snakeyaml.scanner.ScannerImpl.fetchAlias(ScannerImpl.java:918)
        at org.yaml.snakeyaml.scanner.ScannerImpl.fetchMoreTokens(ScannerImpl.java:366)
        at org.yaml.snakeyaml.scanner.ScannerImpl.checkToken(ScannerImpl.java:226)
        at org.yaml.snakeyaml.parser.ParserImpl$ParseBlockMappingKey.produce(ParserImpl.java:557)
        at org.yaml.snakeyaml.parser.ParserImpl.peekEvent(ParserImpl.java:157)
        at org.yaml.snakeyaml.parser.ParserImpl.checkEvent(ParserImpl.java:147)
        at org.yaml.snakeyaml.composer.Composer.composeMappingNode(Composer.java:227)
        at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:154)
        at org.yaml.snakeyaml.composer.Composer.composeSequenceNode(Composer.java:201)
        at org.yaml.snakeyaml.composer.Composer.composeNode(Composer.java:152)
        at org.yaml.snakeyaml.composer.Composer.composeDocument(Composer.java:122)
        at org.yaml.snakeyaml.composer.Composer.getSingleNode(Composer.java:105)
        at org.yaml.snakeyaml.constructor.BaseConstructor.getSingleData(BaseConstructor.java:140)
        at org.yaml.snakeyaml.Yaml.loadFromReader(Yaml.java:524)
        at org.yaml.snakeyaml.Yaml.load(Yaml.java:437)
        at com.thoughtworks.go.domain.materials.git.GitModificationParser.parse(GitModificationParser.java:34)
        at com.thoughtworks.go.domain.materials.git.GitCommand.gitLog(GitCommand.java:133)
        at com.thoughtworks.go.domain.materials.git.GitCommand.modificationsSince(GitCommand.java:116)
        at com.thoughtworks.go.config.materials.git.GitMaterial.modificationsSince(GitMaterial.java:135)
        at com.thoughtworks.go.server.service.materials.GitPoller.modificationsSince(GitPoller.java:35)
        at com.thoughtworks.go.server.service.materials.GitPoller.modificationsSince(GitPoller.java:26)
        at com.thoughtworks.go.server.service.MaterialService.modificationsSince(MaterialService.java:128)
        at com.thoughtworks.go.server.materials.ScmMaterialUpdater.insertLatestOrNewModifications(ScmMaterialUpdater.java:56)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.insertLatestOrNewModifications(MaterialDatabaseUpdater.java:157)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterialWithNewRevisions(MaterialDatabaseUpdater.java:149)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater$2.doInTransaction(MaterialDatabaseUpdater.java:108)
        at com.thoughtworks.go.server.transaction.TransactionCallback.doWithExceptionHandling(TransactionCallback.java:23)
        at com.thoughtworks.go.server.transaction.TransactionTemplate.lambda$executeWithExceptionHandling$2(TransactionTemplate.java:43)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
        at com.thoughtworks.go.server.transaction.TransactionTemplate.executeWithExceptionHandling(TransactionTemplate.java:40)
        at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterial(MaterialDatabaseUpdater.java:105)
        at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:64)
        at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:32)
        at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:83)
        at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:63)
        at java.base/java.lang.Thread.run(Unknown Source)
2020-02-24 10:58:12,053 WARN  [194@MessageListener for ConfigMaterialUpdateListener] ConfigMaterialUpdateListener:70 - [Config Material Update] Cannot update configuration part because material update has failed. Reason: while scanning an alias
 in 'string', line 11, column 5:
        * with leading space in the firs ... 
        ^
expected alphabetic or numeric character, but found  (32)
 in 'string', line 11, column 6:
        * with leading space in the first ... 
         ^
```

##### Code snippets/Screenshots

![image](https://user-images.githubusercontent.com/425325/75147429-8ec1c680-56fd-11ea-80a5-849c6ee65aaf.png)

**Comment by arvindsv** [comment]
@ccmtaylor You're right. It is related to 7752, but it is a duplicate of #7709. There is a possible workaround (hack) mentioned there, or if you're not able to do that, the latest experimental release on https://gocd.org/download should have the change, in case you want to test it (please take a DB backup, if you do, since a downgrade is not possible). Thank you! I'll close this issue for now. Both this and #7752 should be fixed in 20.2.0.

**Comment by ketan** [comment]
Apparently the right way to YAML format a multiple line string beginning with a whitespace is to specify an indent indicator:

```
irb(main):013:0> puts({"foo" => "  foo\nbar"}.to_yaml)
---
foo: |2-
    foo
  bar
=> nil
```
